### DISCUSSION TITLE
`WithBoundArgs` and new `{{component}}` typing

### DISCUSSION THREAD

**Discussion by nwhittaker** [typed-ember/glint#588] [Q&A] [answered]
The migration notes recommend typing `{{component}}` as `ComponentLike<SomeChildComponentSignature>`.[^1]

The `WithBoundArgs` doc wraps the original component typing: `WithBoundArgs<typeof SomeBanner, 'kind'>`.[^2]

I would've thought `WithBoundArgs` would wrap the new typing: `WithBoundArgs<ComponentLike<SomeBannerSignature>, 'kind'>`. Does it matter either way?

[^1]: https://typed-ember.gitbook.io/glint/troubleshooting/migrating#component-typing
[^2]: https://typed-ember.gitbook.io/glint/using-glint/glint-types#withboundargs-and-withboundpositionals

**Comment by dfreeman** [comment, marked as answer]
`WithBoundArgs` is happy to accept any `ComponentLike` type as input, and its behavior should be essentially identical whether you give it `typeof SomeBanner` or `ComponentLike<SomeBannerSignature>`.

The difference is that `typeof SomeBanner` promises much _more_ than just the `ComponentLike` bit of its type—that type includes every other piece of `SomeBanner`'s type, including things like private fields, its inheritance hierarchy, etc.

So if you write `typeof SomeBanner` in your signature, you're including all of those extra bits as part of your public interface. And that's fine if the actual `SomeBanner` value is what you're passing to fulfill that type. As [the migration doc you linked](https://typed-ember.gitbook.io/glint/troubleshooting/migrating#component-typing) points out, though, the `{{component}}` helper _doesn't_ return the original class with all its fine details. The object it gives back is _only_ usable as an opaque component-y value: hence, `ComponentLike`.
